raiden/storage/sqlite.py

```diff
@@ -136,6 +155,124 @@ def get_latest_state_snapshot(self) -> Optional[Tuple[int, Any]]:
+            (state_change_id, ),
+        )
+        serialized = cursor.fetchall()
+
+        result = 0, None
+        if serialized:
+            assert len(serialized) == 1
+            last_applied_state_change_id = serialized[0][0]
+            snapshot_state = self.serializer.deserialize(serialized[0][1])
+            return (last_applied_state_change_id, snapshot_state)
+
+        return result
+
+    def get_latest_event_by_data_field(
+            self,
+            filters: typing.Dict[str, str],
+    ) -> EventRecord:
+        """ Return all state changes filtered by a named field and value."""
+        cursor = self.conn.cursor()
+
+        where_clauses = []
+        args = []
+        for field, value in filters.items():
+            where_clauses.append('json_extract(data, ?)=?')
+            args.append(f'$.{field}')
+            args.append(value)
+
+        cursor.execute(
+            "SELECT identifier, source_statechange_id, data FROM state_events WHERE "
+            f"{' AND '.join(where_clauses)}"
+            "ORDER BY identifier DESC LIMIT 1",
+            args,
+        )
+
+        result = EventRecord(
+            event_identifier=0,
+            state_change_identifier=0,
+            data=None,
+        )
+        try:
+            row = cursor.fetchone()
+            if row:
+                event_id = row[0]
+                state_change_id = row[1]
+                event = self.serializer.deserialize(row[2])
+                result = EventRecord(
+                    event_identifier=event_id,
+                    state_change_identifier=state_change_id,
+                    data=event,
+                )
+        except AttributeError:
+            raise InvalidDBData(
+                'Your local database is corrupt. Bailing ...',
+            )
+
+        return result
+
+    def get_latest_state_change_by_data_field(
+            self,
+            filters: typing.Dict[str, str],
+    ) -> StateChangeRecord:
+        """ Return all state changes filtered by a named field and value."""
+        cursor = self.conn.cursor()
+
+        where_clauses = []
+        args = []
+        for field, value in filters.items():
+            where_clauses.append('json_extract(data, ?)=?')
+            args.append(f'$.{field}')
+            args.append(value)
+
+        cursor.execute(
+            "SELECT identifier, data FROM state_changes WHERE "
+            f"{' AND '.join(where_clauses)}"
+            "ORDER BY identifier DESC LIMIT 1",
+            args,
+        )
+
+        result = StateChangeRecord(state_change_identifier=0, data=None)
+        try:
+            row = cursor.fetchone()
+            if row:
+                state_change_id = row[0]
+                state_change = self.serializer.deserialize(row[1])
+                result = StateChangeRecord(
+                    state_change_identifier=state_change_id,
+                    data=state_change,
+                )
+        except AttributeError:
```

Comment:
when is this exception raised?

Reply:
Although not very well documented (Here)[https://docs.python.org/2/library/sqlite3.html#sqlite3.Cursor.execute] but it seems like it `execute` fails, calling `fetchone` or `fetchall` fails with "Cursor has no attribute called X"